File: apps-microservices/crawler-monitor-backend/src/lib/password.js
/**
 * Password hashing & verification using built-in Node crypto (scrypt).
 *
 * Hash format: "scrypt$<N>$<r>$<p>$<saltHex>$<derivedHex>"
 *   N=16384, r=8, p=1 → ~64 MiB RAM, ~80ms on a modern CPU
 *
 * Why scrypt: built-in (no npm dep), modern KDF, designed against ASIC/GPU.
 *
 * Helper to generate a hash from CLI:
 *   node -e "import('./src/lib/password.js').then(m => m.hashPassword('mypass').then(console.log))"
 */

import { scrypt, randomBytes, timingSafeEqual } from 'crypto';
import { promisify } from 'util';

const scryptAsync = promisify(scrypt);

const N = 16384;
const r = 8;
const p = 1;
const KEY_LEN = 64;
const SALT_LEN = 16;
const PREFIX = 'scrypt';

export async function hashPassword(plain) {
  if (typeof plain !== 'string' || plain.length === 0) {
    throw new Error('Password must be a non-empty string');
  }
  const salt = randomBytes(SALT_LEN);
  const derived = await scryptAsync(plain, salt, KEY_LEN, { N, r, p });
  return `${PREFIX}$${N}$${r}$${p}$${salt.toString('hex')}$${derived.toString('hex')}`;
}

export async function verifyPassword(plain, hash) {
  if (typeof plain !== 'string' || typeof hash !== 'string') return false;
  const parts = hash.split('$');
  if (parts.length !== 6 || parts[0] !== PREFIX) return false;
  const [, nStr, rStr, pStr, saltHex, derivedHex] = parts;
  const Np = Number(nStr); const rp = Number(rStr); const pp = Number(pStr);
  if (!Number.isFinite(Np) || !Number.isFinite(rp) || !Number.isFinite(pp)) return false;
  let salt, expected;
  try {
    salt = Buffer.from(saltHex, 'hex');
    expected = Buffer.from(derivedHex, 'hex');
  } catch { return false; }
  if (expected.length === 0) return false;
  let candidate;
  try {
    candidate = await scryptAsync(plain, salt, expected.length, { N: Np, r: rp, p: pp });
  } catch {
    return false;
  }
  if (candidate.length !== expected.length) return false;
  return timingSafeEqual(candidate, expected);
}

/** Returns true if `s` looks like our hash format (cheap shape check, no crypto). */
export function looksLikeScryptHash(s) {
  return typeof s === 'string' && s.startsWith(PREFIX + '$') && s.split('$').length === 6;
}

File: apps-microservices/crawler-monitor-backend/internal/auth/password/password.go
package password

import (
	"crypto/rand"
	"crypto/subtle"
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"golang.org/x/crypto/bcrypt"
	"golang.org/x/crypto/scrypt"
)

const (
	prefix   = "scrypt"
	defaultN = 16384
	defaultR = 8
	defaultP = 1
	keyLen   = 64
	saltLen  = 16
)

// Hash produces a scrypt hash for new passwords.
func Hash(plain string) (string, error) {
	if plain == "" {
		return "", errors.New("password must be non-empty string")
	}
	salt := make([]byte, saltLen)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}
	derived, err := scrypt.Key([]byte(plain), salt, defaultN, defaultR, defaultP, keyLen)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s$%d$%d$%d$%s$%s",
		prefix, defaultN, defaultR, defaultP,
		hex.EncodeToString(salt), hex.EncodeToString(derived)), nil
}

// Verify checks plain against hash. Supports both scrypt (new) and bcrypt (legacy Node.js) formats.
func Verify(plain, hash string) (bool, error) {
	if plain == "" || hash == "" {
		return false, nil
	}
	// bcrypt hashes start with $2b$ or $2a$ or $2y$ (Node.js bcrypt format)
	if strings.HasPrefix(hash, "$2") {
		err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(plain))
		if err == bcrypt.ErrMismatchedHashAndPassword {
			return false, nil
		}
		return err == nil, err
	}
	// scrypt format: scrypt$N$r$p$saltHex$derivedHex
	parts := strings.Split(hash, "$")
	if len(parts) != 6 || parts[0] != prefix {
		return false, nil
	}
	n, err1 := strconv.Atoi(parts[1])
	r, err2 := strconv.Atoi(parts[2])
	p, err3 := strconv.Atoi(parts[3])
	if err1 != nil || err2 != nil || err3 != nil {
		return false, nil
	}
	salt, err := hex.DecodeString(parts[4])
	if err != nil {
		return false, nil
	}
	expected, err := hex.DecodeString(parts[5])
	if err != nil || len(expected) == 0 {
		return false, nil
	}
	candidate, err := scrypt.Key([]byte(plain), salt, n, r, p, len(expected))
	if err != nil {
		return false, nil
	}
	return subtle.ConstantTimeCompare(candidate, expected) == 1, nil
}

// LooksLikeScryptHash returns true if the hash is in scrypt format.
func LooksLikeScryptHash(s string) bool {
	if !strings.HasPrefix(s, prefix+"$") {
		return false
	}
	return len(strings.Split(s, "$")) == 6
}
